aten/src/ATen/core/ivalue_inl.h

```diff
@@ -338,19 +338,30 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
     callbacks_.emplace_back(std::move(callback));
   }
 
+  /**
+  * Add a callback to the future that that runs with the thread local state of
+  * the caller of this function. This is essentially a wrapper around
+  * at::ThreadLocalStateGuard.
+  */
+  void addCallbackWithTLSState(std::function<void(void)> callback);
+
   /**
    * Add a callback to the future, and return another Future to hold the return
    * value of the callback. This is necessary when the callback provider needs
    * to know for sure when the callback has finished.
    */
   c10::intrusive_ptr<Future> then(
       std::function<IValue(void)> callback,
-      TypePtr type) {
+      TypePtr type, bool propagateTLSState = false) {
     auto fut = c10::make_intrusive<Future>(type);
+    // Pick callback function based on propagateTLSState
+    void (ivalue::Future::*addCallbackFn)(std::function<void(void)>) =
+        propagateTLSState ? &ivalue::Future::addCallbackWithTLSState
```

Comment:
can `propagateTLSState` be a argument in `Future::addCallback` instead of having two methods + manipulating function pointers?

Reply:
Sure, I think that this is reasonable as well.